From adef311c6d1f079d533c861ad157b1fb906bfb2f Mon Sep 17 00:00:00 2001
From: Victor Gu <xigu@marvell.com>
Date: Mon, 20 Mar 2017 16:34:09 +0800
Subject: [PATCH 0918/1345] fix: phy: a3700: revise the USB3 comphy setting
 during power on

commit  7ec7f5d77c6b9c9a3a8490c8a68c9845c2bf754f from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- According to design specification, the transmitter should be set
  to high impedence mode during electrical idle.
  Thus transmitter should detect RX at high impedence mode also,
  and delay is needed to accommodate high impedence off latency.
  Otherwise the USB3 will have detection issue that most of the time
  the USB3 device can not be detected at all, or be detected as USB2
  device sometimes.
  Modified registers:
  RD005C302h (R181h) (0051h) Lane Configuration 1
  Bit 6: set to 1 to let Tx detect Rx at HiZ mode
  Bit [3:4]: set to 2 to be delayed by 2 clock cycles
  Bit 0: set to 1 to set transmitter to high impedance mode during idle.
- USB3 De-emphasize level of -3.5dB is mandatory, but USB3 MAC selects 0x2
  (emphasize disabled) in the MAC_PHY_TXDEEMPH [1:0], while it is supposed
  to select 0x1(3.5dB emphasize). Thus need to override what comes from
  the MAC(by setting register 0x1c2 bit2 to 0x1) and to configure the
  overridded values of MAC_PHY_TXDEEMPH [1:0] to 0x1(bit15 of register
  0x181 and bit0 of register 0x180).
- According to USB3 application note, need to update below comphy
  registers:
  Set max speed generation to USB3.0 5Gbps(set RD005C04Ah bit[11:10] to 1)
  Set capacitor value to 0xF(set RF005C224 bit[3:0] to 0xF)

- JIRA ID: A3700-463 detection issues with USB3.0 devices over USB3.0 HUB

Change-Id: If78fe6d1b776fe37ee99efd45b2f834bf7c98389
Signed-off-by: Victor Gu <xigu@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37641
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/phy/phy-comphy-a3700.c |   36 ++++++++++++++++++++++++++++--------
 drivers/phy/phy-comphy-a3700.h |   16 +++++++++++++++-
 2 files changed, 43 insertions(+), 9 deletions(-)

diff --git a/drivers/phy/phy-comphy-a3700.c b/drivers/phy/phy-comphy-a3700.c
index 40668f5..d631f14 100644
--- a/drivers/phy/phy-comphy-a3700.c
+++ b/drivers/phy/phy-comphy-a3700.c
@@ -528,20 +528,22 @@ static int mvebu_a3700_comphy_usb3_power_on(struct mvebu_comphy_priv *priv,
 	 */
 	usb3_reg_set(reg_base,
 		     COMPHY_REG_LANE_CFG0_ADDR,
-		     PRD_TXDEEMPH_MASK,
-		     (PRD_TXDEEMPH_MASK | PRD_TXMARGIN_MASK |
+		     PRD_TXDEEMPH0_MASK,
+		     (PRD_TXDEEMPH0_MASK | PRD_TXMARGIN_MASK |
 		      PRD_TXSWING_MASK | CFG_TX_ALIGN_POS_MASK),
 		     mode);
 
 	/*
-	 * 2. Unset BIT0: set Tx Electrical Idle Mode
-	 *    unset BIT4: set G2 Tx Datapath with no Delayed Latency
-	 *    unset BIT6: set Tx Detect Rx Mode at LoZ mode
+	 * 2. Set BIT0: enable transmitter in high impedance mode
+	 *    Set BIT[3:4]: delay 2 clock cycles for HiZ off latency
+	 *    Set BIT6: Tx detect Rx at HiZ mode
+	 *    Unset BIT15: set to 0 to set USB3 De-emphasize level to -3.5db
+	 *                 together with bit 0 of COMPHY_REG_LANE_CFG0_ADDR register
 	 */
 	usb3_reg_set(reg_base,
 		     COMPHY_REG_LANE_CFG1_ADDR,
-		     0x0,
-		     REG_16_BIT_MASK,
+		     TX_DET_RX_MODE | GEN2_TX_DATA_DLY_DEFT | TX_ELEC_IDLE_MODE_EN,
+		     PRD_TXDEEMPH1_MASK | TX_DET_RX_MODE | GEN2_TX_DATA_DLY_MASK | TX_ELEC_IDLE_MODE_EN,
 		     mode);
 
 	/*
@@ -665,7 +667,25 @@ static int mvebu_a3700_comphy_usb3_power_on(struct mvebu_comphy_priv *priv,
 			     mode);
 
 	/*
-	 * 14. Release SW reset
+	 * 14. Set max speed generation to USB3.0 5Gbps
+	 */
+	usb3_reg_set(reg_base,
+		     COMPHY_SYNC_MASK_GEN_REG,
+		     PHY_GEN_USB3_5G,
+		     PHY_GEN_MAX_MASK,
+		     mode);
+
+	/*
+	 * 15. Set capacitor value for FFE gain peaking to 0xF
+	 */
+	usb3_reg_set(reg_base,
+		     COMPHY_REG_GEN3_SETTINGS_3,
+		     COMPHY_GEN_FFE_CAP_SEL_VALUE,
+		     COMPHY_GEN_FFE_CAP_SEL_MASK,
+		     mode);
+
+	/*
+	 * 16. Release SW reset
 	 */
 	usb3_reg_set(reg_base,
 		     COMPHY_REG_GLOB_PHY_CTRL0_ADDR,
diff --git a/drivers/phy/phy-comphy-a3700.h b/drivers/phy/phy-comphy-a3700.h
index cf7cc8d..7886aa9 100644
--- a/drivers/phy/phy-comphy-a3700.h
+++ b/drivers/phy/phy-comphy-a3700.h
@@ -90,6 +90,11 @@ enum {
 #define TXD_INVERT_BIT				BIT(10)
 #define RXD_INVERT_BIT				BIT(11)
 
+#define COMPHY_SYNC_MASK_GEN_REG		0x25
+#define PHY_GEN_MAX_OFFSET			10
+#define PHY_GEN_MAX_MASK			(3 << PHY_GEN_MAX_OFFSET)
+#define PHY_GEN_USB3_5G				(1 << PHY_GEN_MAX_OFFSET)
+
 #define COMPHY_ISOLATION_CTRL_REG		0x26
 #define ISOLATION_CTRL_REG_ADDR(unit)		(COMPHY_ISOLATION_CTRL_REG * PHY_SHFT(unit))
 #define PHY_ISOLATE_MODE			BIT(15)
@@ -132,16 +137,25 @@ enum {
 #define MISC_REG1_ADDR(unit)			(COMPHY_MISC_REG1_ADDR * PHY_SHFT(unit))
 #define SEL_BITS_PCIE_FORCE			BIT(15)
 
+#define COMPHY_REG_GEN3_SETTINGS_3		0x112
+#define COMPHY_GEN_FFE_CAP_SEL_MASK		0xF
+#define COMPHY_GEN_FFE_CAP_SEL_VALUE		0xF
+
 #define COMPHY_REG_LANE_CFG0_ADDR		0x180
 #define LANE_CFG0_ADDR(unit)			(COMPHY_REG_LANE_CFG0_ADDR * PHY_SHFT(unit))
-#define PRD_TXDEEMPH_MASK			BIT(0)
+#define PRD_TXDEEMPH0_MASK			BIT(0)
 #define PRD_TXMARGIN_MASK			(BIT(1) | BIT(2) | BIT(3))
 #define PRD_TXSWING_MASK			BIT(4)
 #define CFG_TX_ALIGN_POS_MASK			(BIT(5) | BIT(6) | BIT(7) | BIT(8))
 
 #define COMPHY_REG_LANE_CFG1_ADDR		0x181
 #define LANE_CFG1_ADDR(unit)			(COMPHY_REG_LANE_CFG1_ADDR * PHY_SHFT(unit))
+#define PRD_TXDEEMPH1_MASK			BIT(15)
 #define USE_MAX_PLL_RATE_EN			BIT(9)
+#define TX_DET_RX_MODE				BIT(6)
+#define GEN2_TX_DATA_DLY_MASK			(BIT(3) | BIT(4))
+#define GEN2_TX_DATA_DLY_DEFT			(2 << 3)
+#define TX_ELEC_IDLE_MODE_EN			BIT(0)
 
 #define COMPHY_REG_LANE_STATUS1_ADDR		0x183
 #define LANE_STATUS1_ADDR(unit)			(COMPHY_REG_LANE_STATUS1_ADDR * PHY_SHFT(unit))
-- 
1.7.9.5

